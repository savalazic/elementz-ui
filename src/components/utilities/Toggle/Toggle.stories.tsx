/* tslint:disable:jsx-no-lambda */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { infoConfig } from '../../../utils/info-config';

import { Toggle } from './Toggle';
import { Button } from '../../elements';

const stories = storiesOf('Utilities|Toggle', module);
stories.addParameters(infoConfig);

stories.add('default', () => {
  return (
    <Toggle initialOn={false} onToggle={action('onToggle')}>
      {({ on, reset, toggle }) => (
        <div>
          <Button onClick={() => reset(on)}>Reset</Button>
          <Button type="primary" onClick={() => toggle(on)}>
            Toggle
          </Button>
          <p>{on ? 'ON' : 'OFF'}</p>
        </div>
      )}
    </Toggle>
  );
});
