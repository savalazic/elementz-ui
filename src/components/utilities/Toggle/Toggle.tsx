import React, { Component } from 'react';

export interface ToggleState {
  on: boolean;
}

export interface ToggleProps {
  /** @default off */
  /** initial on */
  initialOn?: boolean;
  /** callback for toggling on/off */
  onToggle?: (on: boolean) => void;
  /** callback on resetting toggle */
  onReset?: (on: boolean) => void;
  /** content */
  children: (
    params: {
      on: boolean;
      toggle: (on: boolean) => void;
      reset: (on: boolean) => void;
    },
  ) => React.ReactNode;
}

class Toggle extends Component<ToggleProps, ToggleState> {
  public initialState = {
    on: this.props.initialOn || false,
  };

  public state = this.initialState;

  public reset = () => {
    this.setState(this.initialState, () => {
      if (this.props.onReset) {
        this.props.onReset(this.state.on);
      }
    });
  };

  public toggle = () => {
    this.setState(
      ({ on }) => ({ on: !on }),
      () => {
        if (this.props.onToggle) {
          this.props.onToggle(this.state.on);
        }
      },
    );
  };

  public render() {
    return this.props.children({
      on: this.state.on,
      toggle: this.toggle,
      reset: this.reset,
    });
  }
}

export { Toggle };
