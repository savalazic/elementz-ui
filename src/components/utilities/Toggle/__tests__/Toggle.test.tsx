import 'jest-dom/extend-expect';
import React from 'react';
import { render } from 'react-testing-library';

import { Toggle } from '../Toggle';

describe('Toggle component', () => {
  it('should match default snapshot', () => {
    const { container } = render(<Toggle />);
    expect(container).toMatchSnapshot();
  });
});
