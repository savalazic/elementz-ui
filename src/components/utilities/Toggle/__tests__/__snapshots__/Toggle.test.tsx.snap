// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle component should match default snapshot 1`] = `
<div>
  <div>
    OFF
  </div>
</div>
`;
